### PR TITLE
Replace systems implementation by Systems.jl

### LINKED ISSUE #41: Refactoring: replace systems implementation by Systems.jl
Possible problem: The discrete system does not store `δ`.

### LINKED ISSUE #121: Accept linear case in `discr_bloat_firstorder`
if `inputdim(cont_sys) == 0` the algorithm `discr_bloat_firstorder` should work by defining the linear particular case.

### LINKED ISSUE #122: Accept time varying case in `discr_bloat_firstorder`

### LINKED ISSUE #143: Option: Sparse/dense matrix conversion
proposal: remove the `try`.

```
        if options[:assume_sparse] && !(Δ.s.A isa SparseMatrixCSC)
            A = sparse(Δ.s.A)
        elseif options[:assume_sparse] && !(Δ.s.A isa Matrix)
            A = full(Δ.s.A)
       else
            A = nothing
       end
       if A != nothing
           Δ = DiscreteSystem(A, Δ.x0, inputset(Δ))
       end
```

### PR DESCRIPTION
**Pull request by mforets**
Closes #41.
Closes #121.
Closes #122.
Closes #143.

### CODE REVIEW

**Comment by mforets** [comment]
this was a branch that i started some time ago that i'd like to get updated. agreed, let's not put delta in the type. to begin with i won't break the current interface, that creates `ContinuousSystem(A, X0)`, `ContinuousSystem(A, X0, U)` and such. we can do that in a second round.

on a separate note i have the impression that the iterator-inputs feature could be a nice add-on to systems. thoughts?

**Comment by schillic** [comment]
>  have the impression that the iterator-inputs feature could be a nice add-on to upstream Systems.

Good idea :+1:

File: src/inputs.jl [hunk 1]
```diff
@@ -203,17 +194,21 @@ struct ContinuousSystem <: AbstractSystem
...
         new(A, X0, U)
 end
 =#
 
-# constructor with no inputs
-ContinuousSystem(A::AbstractMatrix{Float64},
-                 X0::LazySet) =
-    ContinuousSystem(A, X0, ConstantNonDeterministicInput(ZeroSet(size(A, 1))))
+
+# homogeneous and no input
```
Comment (schillic) [inline src/inputs.jl]: One of them suffices, right?

File: src/inputs.jl [hunk 2]
```diff
@@ -50,9 +50,11 @@ struct ConstantNonDeterministicInput <: NonDeterministicInput
     U::LazySet
 end
 
-Base.next(inputs::ConstantNonDeterministicInput, state) = (inputs.U, state + 1)
-Base.done(inputs::ConstantNonDeterministicInput, state) = state > 1
-Base.length(inputs::ConstantNonDeterministicInput) = 1
+const CNDI = ConstantNonDeterministicInput
```
Comment (schillic) [inline src/inputs.jl]: :+1:

**Comment by mforets** [comment]
I changed my mind and suggest that we keep the names `ContinuousSystem` and `DiscreteSystem` for now. If we later extend the functionality of this package to systems other than affine ODEs we can revise this.

**Comment by schillic** [comment]
The build error is because you have to include the systems from `../systems.jl` in `Utils`.

**Comment by mforets** [comment]
👍  I've added it to the `Utils` module. There are still some minor changes before it works on the whole package.

File: src/ReachSets/discretize.jl [hunk 3]
```diff
@@ -78,23 +79,23 @@ This uses a first order approximation of the ODE, and matrix norm upper bounds,
...
 
     if !(cont_sys.U isa ConstantNonDeterministicInput)
         error("This discretization algorithm is only implemented for constant inputs")
     end
 
     Anorm = norm(full(cont_sys.s.A), Inf)
-    RX0 = norm(cons_sys.x0, Inf)
-    inputs = next_set(cont_sys.U)
+    RX0 = norm(X0, Inf)
+    inputs = next_set(inputset(cont_sys))
```
Comment (schillic) [inline src/ReachSets/discretize.jl]: `next_set` is called `nextinput` now, right?

Comment (mforets) [reply src/ReachSets/discretize.jl]: i didn't matter that we kept the name `next_set` inside Reachability and their behavior are slightly different (`nextinput` gives input and state but `next_set` only the first element).

Comment (schillic) [reply src/ReachSets/discretize.jl]: I might be wrong, but I think this function is not defined anymore. (You removed the file `Systems.jl`.)

Comment (mforets) [reply src/ReachSets/discretize.jl]: `next_set` is defined in `Utils/systems.jl`. at the moment, this file exports

```julia
export ContinuousSystem,
       DiscreteSystem,
       add_dimension,
       next_set
```

**Comment by mforets** [comment]
i doubt about writing:

```julia
    if :U in fieldnames(cs.s)
        Uext = map(add_dimension, inputset(cs))
        return ContinuousSystem(Aext, X0ext, Uext)
    else
        return ContinuousSystem(Aext, X0ext)
    end
```

which has to be used because systems without input are instantiated with a `LinearContinuousSystem` (and they don't have an input field).

alternatively that `ContinuousSystem` (and `DiscreteSystem`) always build a `ConstrainedLinearControlContinuousSystem`, and in this case, the matrix `B` will be zero, and `U` will be nothing or `ZeroSet`.

**Comment by mforets** [comment]
> I have no preferences here. Maybe we should just use the types from Systems.jl. If at some point we want to support more types of systems, only ContinuousSystem will not be sufficient.

i do agree to eventually use only types defined in systems. to extend it to nonlinear or hybrid systems, just `ContinuousSystes` won't make sense.

File: src/ReachSets/discretize.jl (hunk 3 above)
Comment (mforets) [reply src/ReachSets/discretize.jl]: in never pushed `systems.jl` :D

**Comment by mforets** [comment]
Finally, green light! I'll stash the commits into a single one, after that i think this is ok.

**Comment by mforets** [comment]
WARNING: something is dead wrong in my changeset, eg. ISS in `mforets/41` takes 22.6s vs. 1.7 in` master`!?

**Comment by schillic** [comment]
> ISS in mforets/41 takes 22.6s vs. 1.7 inmaster!?

This is rather strange because we do not use the systems after the initial discretization phase. Is the bottleneck there or somewhere else?

**Comment by mforets** [comment]
i think the culprit is

```julia
    try
        if options[:assume_sparse] && !(Δ.A isa SparseMatrixCSC)
            A = sparse(Δ.A)
        elseif options[:assume_sparse] && !(Δ.A isa Matrix)
            A = full(Δ.A)
        else
            A = nothing
        end
        if A != nothing
            Δ = DiscreteSystem(A, Δ.X0, Δ.U)
        end
    end
```

here `Δ.A` is meaningless, now it is `Δ.s.A`, but since this is wrapped in a `try`, the error is just ignored.

**Comment by mforets** [comment]
That change fixed it. 

Still, we need [#308](https://github.com/JuliaReach/LazySets.jl/pull/309) for the `lazy_exp` option.

**Comment by mforets** [comment]
I suggest that we run a test suite before merging, to be sure that there are no other bad surprises.

**Comment by schillic** [comment]
> I suggest that we run a test suite before merging, to be sure that there are no other bad surprises.

Did you mean that *I* should do that?

Also, do you see the broken test?

**Comment by mforets** [comment]
> Did you mean that I should do that?

anyone.

> Also, do you see the broken test?

yes :/ this is because Systems types are defined with `AbstractMatrix`. i've opened [#311](https://github.com/JuliaReach/LazySets.jl/issues/311)..

**Comment by mforets** [comment]
there is another problem: by design, the coefficients matrices in `Systems`, of a given system, should be of the same type. since i wrote `eye(A)` for the "B" matrix, this is problematic when `A` is a `SparseMatarixExp`. 

proposal: let's make `eye(A)` return an identity matrix but as a `SparseMatrixExp`.

**Pull request merged**